src/System.Private.Uri/src/System/UriHelper.cs

```diff
@@ -298,6 +298,37 @@ static void EnsureCapacity(char[]? dest, int destSize, int requiredSize)
             }
         }
 
+        internal static unsafe char[] UnescapeString(string input, int start, int end, char[] dest,
+            ref int destPosition, char rsvd1, char rsvd2, char rsvd3, UnescapeMode unescapeMode, UriParser? syntax,
+            bool isQuery)
+        {
+            fixed (char* pStr = input)
+            {
+                return UnescapeString(pStr, start, end, dest, ref destPosition, rsvd1, rsvd2, rsvd3, unescapeMode,
+                    syntax, isQuery);
+            }
+        }
+
+        internal static unsafe char[] UnescapeString(char* pStr, int start, int end, char[] dest, ref int destPosition,
+            char rsvd1, char rsvd2, char rsvd3, UnescapeMode unescapeMode, UriParser? syntax, bool isQuery)
+        {
+            ValueStringBuilder vsb = new ValueStringBuilder(dest.Length);
+            vsb.Append(dest.AsSpan(0, destPosition));
+            UnescapeString(pStr, start, end, ref vsb, rsvd1, rsvd2, rsvd3, unescapeMode,
+                    syntax, isQuery);
+            if (vsb.Length > dest.Length)
+            {
+                dest = vsb.AsSpan().ToArray();
```

Comment:
Do your perf tests measure only `Uri.UnescapeUriString`, or are you also measuring the cost of this as part of Uri processing?  It occurs to me looking at this again that this UnescapeString helper is called (indirectly) from a bunch of places during Uri's processing, which means there may be subsequent processing to be performed that would require writing into the destination array.  Previously, the resizing behavior would likely end up leaving some empty space at the end of the returned array, but now it's exactly sized to fit the output in the case where we need to allocate a new array here.  In theory, that could actually make things worse for some situations, because callers may need to end up reallocating in places they wouldn't have had to before.

I don't think we should hold up this change for that, but it's something to think about subsequently.  I think it's ok to go ahead with it though because we should look to replace all of the array based code with VSB, rather than going back and forth between the two, so the issue would largely go away.

Reply:
I measured perf of Uri.UnescapeDataString only. I see your point here but it seems to be tricky to design good perf tests matching real-world scenarios for the mentioned cases because it's not clear how often the dest's tail is currently used in the subsequent processing.

Reply:
Various operations performed on commonly shaped Uris would be a good start.